// mqtts/dist/transport/tcp.transport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const transport_1 = require("./transport");
const URL = require("url");
const net_1 = require("net");
class TcpTransport extends transport_1.Transport {
    send(data) {
        this.socket.write(data);
    }
    connect() {
        var _a;
        const url = URL.parse(this.options.url);
        this.socket = net_1.connect({
            host: (_a = url.hostname, (_a !== null && _a !== void 0 ? _a : '')),
            port: Number(url.port),
            timeout: 0,
        });
        this.socket.on('error', e => this.callbacks.error(e));
        this.socket.on('end', () => this.callbacks.disconnect());
        this.socket.on('close', () => this.callbacks.disconnect());
        this.socket.on('connect', () => this.callbacks.connect());
        this.socket.on('timeout', () => this.callbacks.disconnect());
        this.socket.on('data', res => this.callbacks.data(res));
    }
    disconnect() {
        this.socket.removeAllListeners('close');
        this.socket.end();
    }
}
exports.TcpTransport = TcpTransport;
