// mqtts/dist/packet-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/*

    Based on: https://github.com/binsoul/net-mqtt/blob/master/src/PacketStream.php
    Last revision: 10-Sep-19

 */
const errors_1 = require("./errors");
class PacketStream {
    constructor(data, length, buffer) {
        this._data = data ? Buffer.from(data) : length ? Buffer.alloc(length) : buffer ? buffer : Buffer.from([]);
        this.position = 0;
    }
    set position(value) {
        this._position = value;
    }
    get position() {
        return this._position;
    }
    get data() {
        return this._data;
    }
    get length() {
        return this._data ? this._data.length : 0;
    }
    get remainingBytes() {
        return this.length - this.position;
    }
    static fromLength(len) {
        return new PacketStream(undefined, len, undefined);
    }
    static fromBuffer(buf) {
        return new PacketStream(undefined, undefined, buf);
    }
    static fromString(data) {
        return new PacketStream(data, undefined, undefined);
    }
    static empty() {
        return new PacketStream(undefined, undefined, undefined);
    }
    /**
     *
     * @param {number} steps - steps to move
     * @returns {number} Position before moving
     */
    move(steps = 1) {
        this._position += steps;
        if (this._position > this.length)
            throw new errors_1.EndOfStreamError('Reached end of stream');
        return this._position - steps;
    }
    // General
    seek(len) {
        this.move(len);
        return this;
    }
    cut() {
        this._data = this._data.slice(this._position) || undefined;
        this._position = 0;
        return this;
    }
    // Write
    write(data) {
        if (this._data)
            this._data = Buffer.concat([this._data, data]);
        else
            this._data = data;
        this.move(data.length);
        return this;
    }
    writeRawString(data) {
        return this.write(Buffer.from(data));
    }
    writeByte(num) {
        this.write(Buffer.from([num]));
        return this;
    }
    writeWord(num) {
        return this.write(Buffer.from([(num & 0xff00) >> 8, num & 0xff]));
    }
    writeString(str) {
        this.writeWord(Buffer.byteLength(str));
        return this.writeRawString(str);
    }
    // Read
    read(len) {
        if (this.position > this.length || len > this.length - this.position) {
            throw new errors_1.EndOfStreamError(`End of stream reached when trying to read ${len} bytes. content length=${this.length}, position=${this.position}`);
        }
        const buf = this._data.slice(this._position, this.position + len);
        this.move(len);
        return buf;
    }
    readSlice(end) {
        const buf = this._data.slice(this._position, end);
        this.move(buf.length);
        return buf;
    }
    readByte() {
        return this._data.readUInt8(this.move(1));
    }
    readWord() {
        return this._data.readUInt16BE(this.move(2));
    }
    readString() {
        const len = this.readWord();
        return this.read(len).toString('utf8');
    }
    readStringAsBuffer() {
        return this.read(this.readWord());
    }
}
exports.PacketStream = PacketStream;
