// instagram_mqtt/dist/realtime/subscriptions/graphql.subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLSubscriptions = exports.QueryIDs = void 0;
const Chance = require("chance");
// src: com.instagram.realtimeclient.RealtimeSubscription
exports.QueryIDs = {
    appPresence: '17846944882223835',
    asyncAdSub: '17911191835112000',
    clientConfigUpdate: '17849856529644700',
    directStatus: '17854499065530643',
    directTyping: '17867973967082385',
    liveWave: '17882305414154951',
    interactivityActivateQuestion: '18005526940184517',
    interactivityRealtimeQuestionSubmissionsStatus: '18027779584026952',
    interactivitySub: '17907616480241689',
    liveRealtimeComments: '17855344750227125',
    liveTypingIndicator: '17926314067024917',
    mediaFeedback: '17877917527113814',
    reactNativeOTA: '17861494672288167',
    videoCallCoWatchControl: '17878679623388956',
    videoCallInAlert: '17878679623388956',
    videoCallPrototypePublish: '18031704190010162',
    zeroProvision: '17913953740109069',
};
class GraphQLSubscriptions {
    static formatSubscriptionString(queryId, inputParams, clientLogged) {
        return `1/graphqlsubscriptions/${queryId}/${JSON.stringify({
            input_data: inputParams,
            ...(clientLogged ? { '%options': { client_logged: clientLogged } } : {}),
        })}`;
    }
}
exports.GraphQLSubscriptions = GraphQLSubscriptions;
GraphQLSubscriptions.getAppPresenceSubscription = (options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.appPresence, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
}, options.clientLogged);
GraphQLSubscriptions.getAsyncAdSubscription = (userId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.asyncAdSub, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    user_id: userId,
}, options.clientLogged);
GraphQLSubscriptions.getClientConfigUpdateSubscription = (options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.clientConfigUpdate, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
}, options.clientLogged);
GraphQLSubscriptions.getDirectStatusSubscription = (options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.directStatus, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
}, options.clientLogged);
GraphQLSubscriptions.getDirectTypingSubscription = (userId, clientLogged) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.directTyping, {
    user_id: userId,
}, clientLogged);
GraphQLSubscriptions.getIgLiveWaveSubscription = (broadcastId, receiverId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.liveWave, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
    receiver_id: receiverId,
}, options.clientLogged);
GraphQLSubscriptions.getInteractivityActivateQuestionSubscription = (broadcastId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.interactivityActivateQuestion, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
}, options.clientLogged);
GraphQLSubscriptions.getInteractivityRealtimeQuestionSubmissionsStatusSubscription = (broadcastId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.interactivityRealtimeQuestionSubmissionsStatus, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
}, options.clientLogged);
GraphQLSubscriptions.getInteractivitySubscription = (broadcastId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.interactivitySub, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
}, options.clientLogged);
GraphQLSubscriptions.getLiveRealtimeCommentsSubscription = (broadcastId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.liveRealtimeComments, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
}, options.clientLogged);
GraphQLSubscriptions.getLiveTypingIndicatorSubscription = (broadcastId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.liveTypingIndicator, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    broadcast_id: broadcastId,
}, options.clientLogged);
GraphQLSubscriptions.getMediaFeedbackSubscription = (feedbackId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.mediaFeedback, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    feedback_id: feedbackId,
}, options.clientLogged);
GraphQLSubscriptions.getReactNativeOTAUpdateSubscription = (buildNumber, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.reactNativeOTA, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    build_number: buildNumber,
}, options.clientLogged);
GraphQLSubscriptions.getVideoCallCoWatchControlSubscription = (videoCallId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.videoCallCoWatchControl, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    video_call_id: videoCallId,
}, options.clientLogged);
GraphQLSubscriptions.getVideoCallInCallAlertSubscription = (videoCallId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.videoCallInAlert, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    video_call_id: videoCallId,
}, options.clientLogged);
GraphQLSubscriptions.getVideoCallPrototypePublishSubscription = (videoCallId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.videoCallPrototypePublish, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    video_call_id: videoCallId,
}, options.clientLogged);
GraphQLSubscriptions.getZeroProvisionSubscription = (deviceId, options = {}) => GraphQLSubscriptions.formatSubscriptionString(exports.QueryIDs.zeroProvision, {
    client_subscription_id: options.subscriptionId || new Chance().guid({ version: 4 }),
    device_id: deviceId,
}, options.clientLogged);
