// instagram_mqtt/dist/realtime/messages/message-sync.message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageSyncMessageTypes = void 0;
var MessageSyncMessageTypes;
(function (MessageSyncMessageTypes) {
    MessageSyncMessageTypes["DELETION"] = "deletion";
    MessageSyncMessageTypes["MEDIA"] = "media";
    MessageSyncMessageTypes["TEXT"] = "text";
    MessageSyncMessageTypes["LIKE"] = "like";
    MessageSyncMessageTypes["HASHTAG"] = "hashtag";
    MessageSyncMessageTypes["PROFILE"] = "profile";
    MessageSyncMessageTypes["MEDIA_SHARE"] = "media_share";
    MessageSyncMessageTypes["LOCATION"] = "location";
    MessageSyncMessageTypes["ACTION_LOG"] = "action_log";
    MessageSyncMessageTypes["TITLE"] = "title";
    MessageSyncMessageTypes["USER_REACTION"] = "user_reaction";
    MessageSyncMessageTypes["HISTORY_EDIT"] = "history_edit";
    MessageSyncMessageTypes["REACTION_LOG"] = "reaction_log";
    MessageSyncMessageTypes["REEL_SHARE"] = "reel_share";
    MessageSyncMessageTypes["DEPRECATED_CHANNEL"] = "deprecated_channel";
    MessageSyncMessageTypes["LINK"] = "link";
    MessageSyncMessageTypes["RAVEN_MEDIA"] = "raven_media";
    MessageSyncMessageTypes["LIVE_VIDEO_SHARE"] = "live_video_share";
    MessageSyncMessageTypes["TEST"] = "test";
    MessageSyncMessageTypes["STORY_SHARE"] = "story_share";
    MessageSyncMessageTypes["REEL_REACT"] = "reel_react";
    MessageSyncMessageTypes["LIVE_INVITE_GUEST"] = "live_invite_guest";
    MessageSyncMessageTypes["LIVE_VIEWER_INVITE"] = "live_viewer_invite";
    MessageSyncMessageTypes["TYPE_MAX"] = "type_max";
    MessageSyncMessageTypes["PLACEHOLDER"] = "placeholder";
    MessageSyncMessageTypes["PRODUCT"] = "product";
    MessageSyncMessageTypes["PRODUCT_SHARE"] = "product_share";
    MessageSyncMessageTypes["VIDEO_CALL_EVENT"] = "video_call_event";
    MessageSyncMessageTypes["POLL_VOTE"] = "poll_vote";
    MessageSyncMessageTypes["FELIX_SHARE"] = "felix_share";
    MessageSyncMessageTypes["ANIMATED_MEDIA"] = "animated_media";
    MessageSyncMessageTypes["CTA_LINK"] = "cta_link";
    MessageSyncMessageTypes["VOICE_MEDIA"] = "voice_media";
    MessageSyncMessageTypes["STATIC_STICKER"] = "static_sticker";
    MessageSyncMessageTypes["AR_EFFECT"] = "ar_effect";
    MessageSyncMessageTypes["SELFIE_STICKER"] = "selfie_sticker";
})(MessageSyncMessageTypes = exports.MessageSyncMessageTypes || (exports.MessageSyncMessageTypes = {}));
